src/daft-local-execution/src/sinks/grouped_aggregate.rs

```diff
@@ -0,0 +1,403 @@
+use std::sync::Arc;
+
+use common_error::DaftResult;
+use common_runtime::RuntimeRef;
+use daft_core::prelude::SchemaRef;
+use daft_dsl::{col, Expr, ExprRef};
+use daft_micropartition::MicroPartition;
+use daft_physical_plan::extract_agg_expr;
+use tracing::{info_span, instrument, Instrument};
+
+use super::blocking_sink::{
+    BlockingSink, BlockingSinkFinalizeResult, BlockingSinkSinkResult, BlockingSinkState,
+    BlockingSinkStatus,
+};
+use crate::NUM_CPUS;
+
+enum AggStrategy {
+    // TODO: This would probably benefit from doing sharded aggs.
+    AggThenPartition,
+    PartitionThenAgg,
+    PartitionOnly,
+}
+
+impl AggStrategy {
+    fn execute_strategy(
+        &self,
+        inner_states: &mut [Option<SinglePartitionAggregateState>],
+        input: Arc<MicroPartition>,
+        params: &GroupedAggregateParams,
+    ) -> DaftResult<()> {
+        match self {
+            Self::AggThenPartition => Self::execute_agg_then_partition(inner_states, input, params),
+            Self::PartitionThenAgg => Self::execute_partition_then_agg(inner_states, input, params),
+            Self::PartitionOnly => Self::execute_partition_only(inner_states, input, params),
+        }
+    }
+
+    fn execute_agg_then_partition(
+        inner_states: &mut [Option<SinglePartitionAggregateState>],
+        input: Arc<MicroPartition>,
+        params: &GroupedAggregateParams,
+    ) -> DaftResult<()> {
+        let agged = input.agg(
+            params.partial_agg_exprs.as_slice(),
+            params.group_by.as_slice(),
+        )?;
+        let partitioned =
+            agged.partition_by_hash(params.final_group_by.as_slice(), inner_states.len())?;
+        for (p, state) in partitioned.into_iter().zip(inner_states.iter_mut()) {
+            let state = state.get_or_insert_with(|| SinglePartitionAggregateState {
```

Comment:
I think it would be cleaner to implement `Default for SinglePartitionAggregateState` then instead you could do 

`state.get_or_insert_default()`